stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1056 @@
+        miner_payouts: Option<&(MinerReward, Vec<MinerReward>, MinerReward, MinerRewardInfo)>,
+    ) -> Result<Vec<StacksTransactionEvent>, ChainstateError> {
+        // add miner payments
+        if let Some((ref miner_reward, ref user_rewards, ref parent_reward, _)) = miner_payouts {
+            // grant in order by miner, then users
+            let matured_ustx = StacksChainState::process_matured_miner_rewards(
+                clarity_tx,
+                miner_reward,
+                user_rewards,
+                parent_reward,
+            )?;
+
+            clarity_tx.increment_ustx_liquid_supply(matured_ustx);
+        }
+
+        // process unlocks
+        let (new_unlocked_ustx, lockup_events) = StacksChainState::process_stx_unlocks(clarity_tx)?;
+
+        clarity_tx.increment_ustx_liquid_supply(new_unlocked_ustx);
+
+        Ok(lockup_events)
+    }
+
+    /// Called in both follower and miner block assembly paths.
+    ///
+    /// Returns clarity_tx, list of receipts, microblock execution cost,
+    /// microblock fees, microblock burns, list of microblock tx receipts,
+    /// miner rewards tuples, the stacks epoch id, and a boolean that
+    /// represents whether the epoch transition has been applied.
+    ///
+    /// The `burn_dbconn`, `sortition_dbconn`, and `conn` arguments
+    ///  all reference the same sortition database through different
+    ///  interfaces. `burn_dbconn` and `sortition_dbconn` should
+    ///  reference the same object. The reason to provide both is that
+    ///  `SortitionDBRef` captures trait functions that Clarity does
+    ///  not need, and Rust does not support trait upcasting (even
+    ///  though it would theoretically be safe).
+    pub fn setup_block<'a, 'b>(
+        chainstate_tx: &'b mut ChainstateTx,
+        clarity_instance: &'a mut ClarityInstance,
+        burn_dbconn: &'b dyn BurnStateDB,
+        sortition_dbconn: &'b dyn SortitionDBRef,
+        conn: &Connection, // connection to the sortition DB
+        pox_constants: &PoxConstants,
+        chain_tip: &StacksHeaderInfo,
+        burn_view: BurnchainHeaderHash,
+        burn_view_height: u32,
+        parent_consensus_hash: ConsensusHash,
+        parent_header_hash: BlockHeaderHash,
+        mainnet: bool,
+        miner_id_opt: Option<usize>,
+        tenure_changed: bool,
+    ) -> Result<SetupBlockResult<'a, 'b>, ChainstateError> {
+        let parent_index_hash = StacksBlockId::new(&parent_consensus_hash, &parent_header_hash);
+        let parent_sortition_id = burn_dbconn
+            .get_sortition_id_from_consensus_hash(&parent_consensus_hash)
+            .expect("Failed to get parent SortitionID from ConsensusHash");
+
+        // find matured miner rewards, so we can grant them within the Clarity DB tx.
+        // TODO: this must be updated to either:
+        //   (A) use the TENURE HEIGHT -- i.e., count of tenures, rather than count of stacks blocks
+        //   (B) use BURNCHAIN HEIGHT
```

Comment:
In the Nakamoto v1 design doc, these are on 1-to-1 correspondence.  Stackers still create a `TenureChange` transaction for each Bitcoin block, even if there was no miner for the tenure.  A subsequent miner just packages the `TenureChange` transactions in order (see Figure 3 here: https://github.com/stacks-network/stacks-blockchain/discussions/3892).

Reply:
I would use tenure height, since then you can correlate it to one of these `TenureChange` transactions and the block that contained it (and then recover the burnchain block in which the Stacks block was mined)

Reply:
Yep -- I agree, tenure height is the thing to use here.

Reply:
Implemented miner reward schedules in this PR (with assertions in the unit tests) that uses tenure-height to track maturity. It also computes tx fees for the prior tenure at tenure changes.